presto-main/src/main/java/com/facebook/presto/application/ApplicationManager.java

```diff
@@ -0,0 +1,114 @@
+import static com.google.common.base.Preconditions.checkState;
+import static java.lang.String.format;
+import static java.util.Objects.requireNonNull;
+
+@ThreadSafe
+public class ApplicationManager
+{
+    private static final Logger log = Logger.get(ApplicationManager.class);
+
+    @GuardedBy("this")
+    private final Map<String, ApplicationFactory> applicationFactories = new ConcurrentHashMap<>();
+    @GuardedBy("this")
+    private final Map<String, Application> applications = new ConcurrentHashMap<>();
+
+    private final AtomicBoolean stopped = new AtomicBoolean();
+
+    public synchronized void addApplicationFactory(ApplicationFactory applicationFactory)
+    {
+        requireNonNull(applicationFactory, "applicationFactory is null");
+
+        if (applicationFactories.putIfAbsent(applicationFactory.getName(), applicationFactory) != null) {
+            throw new IllegalArgumentException(format("Application '%s' is already registered", applicationFactory.getName()));
+        }
+    }
+
+    @PreDestroy
+    public void stop()
+    {
+        if (stopped.getAndSet(true)) {
+            return;
+        }
+
+        for (Map.Entry<String, Application> entry : applications.entrySet()) {
+            Application application = entry.getValue();
+            try (ThreadContextClassLoader ignored = new ThreadContextClassLoader(application.getClass().getClassLoader())) {
+                application.shutdown();
+            }
+            catch (Throwable t) {
+                log.error(t, "Error shutting down application: %s", entry.getKey());
+            }
+        }
+    }
+
+    protected synchronized void setup(String applicationName, ApplicationFactory applicationFactory, Map<String, String> properties)
+    {
+        checkState(!stopped.get(), "ApplicationManager is stopped");
+        checkState(!applications.containsKey(applicationName), "Application %s already exists", applicationName);
+        try {
+            applications.put(applicationName, applicationFactory.create(properties));
+        }
+        catch (Throwable t) {
+            log.error(t, "Error initializing application: %s", applicationFactory.getName());
+        }
+    }
+
+    public void setup(String applicationName, Map<String, String> properties)
+    {
+        checkState(!stopped.get(), "ApplicationManager is stopped");
+        ApplicationFactory applicationFactory = applicationFactories.get(applicationName);
+        checkArgument(applicationFactory != null, "No factory for application %s", applicationName);
+        setup(applicationName, applicationFactory, properties);
+    }
+
+    protected void start(Injector injector)
+    {
+        for (Map.Entry<String, Application> entry : applications.entrySet()) {
+            Application application = entry.getValue();
+            try (ThreadContextClassLoader ignored = new ThreadContextClassLoader(application.getClass().getClassLoader())) {
+                application.run(injector);
```

Comment:
Initially, I suppose to bind UI on the port other than the API port here: #7106. In that PR, I just add a new module inside the PrestoServer and rebind the instance in the initialized injector to another Bootstrap, which looks hacky. I feel it would be reasonable adding an application plugin which runs on top of presto-main, and potentially passing some context instances from the initialized injector in some way. I agree that it shouldn't pass the whole injector and it shouldn't expose all internal instance to plugins.